src/CabanaPD_ForceModelsMulti.hpp

```diff
@@ -17,41 +17,93 @@
 namespace CabanaPD
 {
 
-/******************************************************************************
-  Multi-material models
-******************************************************************************/
-struct AverageTag
+namespace Impl
+{
+template <std::size_t current, typename FunctorType, typename ParameterPackType,
+          typename... ARGS,
+          std::enable_if_t<current<ParameterPackType::size - 1, int> = 0>
+              KOKKOS_INLINE_FUNCTION auto
+                  recursion_functor_for_index_in_pack_with_args(
+                      FunctorType const& functor, std::size_t index,
+                      ParameterPackType& parameterPack, ARGS... args )
+{
+    if ( index == current )
+    {
+        return functor( Cabana::get<current>( parameterPack ), args... );
+    }
+    else
+    {
+        return recursion_functor_for_index_in_pack_with_args<current + 1>(
+            functor, index, parameterPack, args... );
+    }
+}
+
+template <std::size_t current, typename FunctorType, typename ParameterPackType,
+          typename... ARGS,
+          std::enable_if_t<current == ParameterPackType::size - 1, int> = 0>
+KOKKOS_INLINE_FUNCTION auto recursion_functor_for_index_in_pack_with_args(
+    FunctorType const& functor, std::size_t index,
+    ParameterPackType& parameterPack, ARGS... args )
+{
+    if ( index == current )
+    {
+        return functor( Cabana::get<current>( parameterPack ), args... );
+    }
+    else
+    {
+        Kokkos::abort( "Requested index not contained in ParameterPack" );
+        return functor( Cabana::get<current>( parameterPack ), args... );
+    }
+}
+
+template <typename FunctorType, typename ParameterPackType, typename... ARGS>
+KOKKOS_INLINE_FUNCTION auto run_functor_for_index_in_pack_with_args(
+    FunctorType const& functor, std::size_t index,
+    ParameterPackType& parameterPack, ARGS... args )
+{
+    return recursion_functor_for_index_in_pack_with_args<0>(
+        functor, index, parameterPack, args... );
+}
+
+struct IdentityFunctor
 {
+    template <typename Model, typename... ARGS>
+    KOKKOS_FUNCTION auto operator()( Model& model, ARGS... args ) const
+    {
+        return model( args... );
+    }
 };
 
 // Wrap multiple models in a single object.
-// TODO: this currently only supports bi-material systems.
-template <typename MaterialType, typename ModelType1, typename ModelType2,
-          typename ModelType12>
-struct ForceModels
+template <typename MaterialType, typename ParameterPackType, typename Sequence>
+struct ForceModelsImpl
+{
+};
+
+template <typename MaterialType, typename ParameterPackType,
+          std::size_t... Indices>
+struct ForceModelsImpl<MaterialType, ParameterPackType,
+                       std::index_sequence<Indices...>>
 {
     using material_type = MultiMaterial;
 
-    using first_model = ModelType1;
+    // TODO how to decide this?
```

Comment:
I think we discussed, but here we need `base_model` to all match, `thermal_type` to be `TemperatureDependent` if any model is and `fracture_type` to be `Fracture` if any model is. `model_type` can be any which matches `fracture_type`

Reply:
This is implemented in https://github.com/ORNL/CabanaPD/pull/262/commits/3609c85fbfe126fd9fd6d2de1f73aa534cfbd5f7